scripts/make_luals_meta.py

```diff
@@ -242,13 +242,21 @@ def read_file(path: Path, out: TextIOWrapper):
                 )
             name = header[0].split(" ", 1)[1]
             printmsg(path, reader.line_no(), f"enum {name}")
-            out.write(f"---@alias {name} integer\n")
+            variants = reader.read_enum_variants()
+
+            vtypes = []
+            for variant in variants:
+                vtype = f'{name}.{variant}'
+                vtypes.append(vtype)
+                out.write(f'---@alias {vtype} "{vtype}"\n')
+
+            out.write(f"---@alias {name} {'|'.join(vtypes)}\n")
```

Comment:
Why is this done? `---@alias c2.LogLevel.Debug "c2.LogLevel.Debug"` seems wrong. It defines a type `c2.LogLevel.Debug` to be the literal string `c2.LogLevel.Debug`.

Reply:
Yes, this is so the language server doesn't go `number == number` and instead will show the right overload. I'm pretty sure this works.

Reply:
This doesn't work: 

![](https://github.com/user-attachments/assets/6747cb11-70b7-49cd-b59b-2cb6f4cc617b)

Seems like what we really want is [`---@enum <name>`](https://luals.github.io/wiki/annotations/#enum).

<details><summary>Diff</summary>

```diff
diff --git a/scripts/make_luals_meta.py b/scripts/make_luals_meta.py
index b1420e780..cb7b3cdf4 100644
--- a/scripts/make_luals_meta.py
+++ b/scripts/make_luals_meta.py
@@ -242,25 +242,16 @@ def read_file(path: Path, out: TextIOWrapper):
                 )
             name = header[0].split(" ", 1)[1]
             printmsg(path, reader.line_no(), f"enum {name}")
-            variants = reader.read_enum_variants()
-
-            vtypes = []
-            for variant in variants:
-                vtype = f'{name}.{variant}'
-                vtypes.append(vtype)
-                out.write(f'---@alias {vtype} "{vtype}"\n')
-
-            out.write(f"---@alias {name} {'|'.join(vtypes)}\n")
             if header_comment:
                 out.write(f"--- {header_comment}\n")
-            out.write("---@type { ")
+            out.write(f"---@enum {name}\n")
+            out.write(f"{name} = {{\n")
             out.write(
-                ", ".join(
-                    [f"{variant}: {typ}" for variant, typ in zip(variants,vtypes)]
+                "\n".join(
+                    [f"    {variant} = 0," for variant in reader.read_enum_variants()]
                 )
             )
-            out.write(" }\n")
-            out.write(f"{name} = {{}}\n\n")
+            out.write("\n}\n\n")
             continue
 
         # class
```
</details> 

Regardless, this isn't related to the SOL migration.

Reply:
The patch you've sent makes luals fail to typecheck a simple example:
```lua
c2.register_callback(c2.EventType.CompletionRequested, function(ev)end)
--[[
Cannot assign `integer` to parameter `c2.EventType.CompletionRequested`.
- `integer` cannot match `c2.EventType.CompletionRequested`
- Type `integer` cannot match `c2.EventType.CompletionRequested`
- Type `number` cannot match `c2.EventType.CompletionRequested`",
]]
```
For me it also suggests using raw enum values which are all zero which feels even worse than getting strings.

Reply:
Oops, I thought we used numbers for enums. Generating strings here shouldn't be too hard, though.

Reply:
We actually currently can pick whichever representation suits us, the documentation states that the values may be changed any time and accessing through the enum table is the recommended option.